test/standalone/map_xml_test.cpp

```diff
@@ -0,0 +1,206 @@
+                     std::istream_iterator<char>(in_map2)));
+}
+
+// save a map to an XML string, load it again, save it again,
+// then check that it is exactly the same. this is basically the
+// same as the file-based one above, except it exercises the
+// in-memory string functions instead.
+void round_trip_str(mapnik::Map const &m, std::string const &base) {
+  tmp_dir dir;
+
+  std::string test_map1;
+  REQUIRE_NOTHROW(test_map1 = mapnik::save_map_to_string(m));
+
+  mapnik::Map new_map(256, 256);
+  REQUIRE(new_map.register_fonts("fonts", true));
+  REQUIRE_NOTHROW(mapnik::load_map_string(new_map, test_map1, false, base));
+  std::string test_map2;
+  REQUIRE_NOTHROW(test_map2 = mapnik::save_map_to_string(new_map));
+
+  // strings should be identically equal
+  REQUIRE(test_map1 == test_map2);
+}
+
+void compare_map(bfs::path xml) {
+  mapnik::Map m(256, 256);
+  REQUIRE(m.register_fonts("fonts", true));
+  bfs::path abs_base = xml.parent_path();
+
+  // if the attempt to load the map fails, then skip the test.
+  // the "good/bad" testing covers the case of what to do, so
+  // we don't need to test it again here.
+  if (load_map(m, xml, false, abs_base.native())) {
+    round_trip_str(m, abs_base.native());
+    round_trip_file(m, abs_base.native());
+  }
+}
+
+void add_xml_files(bfs::path dir, std::vector<bfs::path> &xml_files) {
+  for (auto const &entry : boost::make_iterator_range(
+         bfs::directory_iterator(dir), bfs::directory_iterator())) {
+    auto path = entry.path();
+    if (path.extension().native() == ".xml") {
+      xml_files.emplace_back(path);
+    }
+  }
+}
+
+} // anonymous namespace
+
+const bool registered = mapnik::datasource_cache::instance().register_datasources("./plugins/input/");
+
+TEST_CASE("map xml I/O") {
+  // make sure plugins are loaded
+  REQUIRE(registered);
+
+  // make the tests silent since we intentially test error conditions that are noisy
+  auto const severity = mapnik::logger::instance().get_severity();
+  mapnik::logger::instance().set_severity(mapnik::logger::none);
+
+  SECTION("good maps") {
+    std::vector<bfs::path> good_maps;
+    add_xml_files("test/data/good_maps", good_maps);
+
+    for (auto const &path : good_maps) {
+      CAPTURE(path.native());
+
+      // check that it can load
+      mapnik::Map m(256, 256);
+      REQUIRE(m.register_fonts("fonts", true));
+      REQUIRE_NOTHROW(load_map(m, path));
+
+      // and, if it can, that it saves the same way
+      compare_map(path);
+    }
+  } // END SECTION
+
+  SECTION("broken maps") {
+    std::vector<bfs::path> broken_maps;
+    add_xml_files("test/data/broken_maps", broken_maps);
+    broken_maps.emplace_back("test/data/broken_maps/does_not_exist.xml");
+
+    for (auto const &path : broken_maps) {
+      CAPTURE(path.native());
+
+      mapnik::Map m(256, 256);
+      REQUIRE(m.register_fonts("fonts", true));
+      REQUIRE_THROWS(mapnik::load_map(m, path.native(), true));
```

Comment:
It would be pretty easy to trap the exceptions in `load_map` - and I'm not sure whether the extra information about the type of error is useful to the caller of `load_map` beyond that it's a config error. @springmeyer, do you think it would be good to have a single, documented expected exception type thrown from `load_map`?